src/polyfill/RTCDataChannel.ts

```diff
@@ -1,57 +1,60 @@
     #dataChannel: DataChannel;
     #readyState: RTCDataChannelState;
-    #bufferedAmountLowThreshold: number;
-    #binaryType: BinaryType;
+    #bufferedAmountLowThreshold: number = 0;
+    #binaryType: BinaryType = 'blob';
     #maxPacketLifeTime: number | null;
     #maxRetransmits: number | null;
     #negotiated: boolean;
     #ordered: boolean;
-
-    #closeRequested = false;
+    #pc: RTCPeerConnection;
 
     // events
-    onbufferedamountlow: ((this: RTCDataChannel, ev: Event) => any) | null;
-    onclose: ((this: RTCDataChannel, ev: Event) => any) | null;
-    onclosing: ((this: RTCDataChannel, ev: Event) => any) | null;
-    onerror: ((this: RTCDataChannel, ev: Event) => any) | null;
-    onmessage: ((this: RTCDataChannel, ev: MessageEvent) => any) | null;
-    onopen: ((this: RTCDataChannel, ev: Event) => any) | null;
-
-    constructor(dataChannel: DataChannel, opts: globalThis.RTCDataChannelInit = {}) {
+    onbufferedamountlow: globalThis.RTCDataChannel['onbufferedamountlow'];
+    onclose: globalThis.RTCDataChannel['onclose'];
+    onclosing: globalThis.RTCDataChannel['onclosing'];
+    onerror: globalThis.RTCDataChannel['onerror'];
+    onmessage: globalThis.RTCDataChannel['onmessage'];
+    onopen: globalThis.RTCDataChannel['onopen']
+
+    constructor(dataChannel: DataChannel, opts: globalThis.RTCDataChannelInit = {}, pc: RTCPeerConnection) {
         super();
 
         this.#dataChannel = dataChannel;
-        this.#binaryType = 'blob';
         this.#readyState = this.#dataChannel.isOpen() ? 'open' : 'connecting';
-        this.#bufferedAmountLowThreshold = 0;
-        this.#maxPacketLifeTime = opts.maxPacketLifeTime || null;
-        this.#maxRetransmits = opts.maxRetransmits || null;
-        this.#negotiated = opts.negotiated || false;
-        this.#ordered = opts.ordered || true;
+        this.#maxPacketLifeTime = opts.maxPacketLifeTime ?? null;
+        this.#maxRetransmits = opts.maxRetransmits ?? null;
+        this.#negotiated = opts.negotiated ?? false;
+        this.#ordered = opts.ordered ?? true;
+        this.#pc = pc
 
         // forward dataChannel events
         this.#dataChannel.onOpen(() => {
             this.#readyState = 'open';
             this.dispatchEvent(new Event('open', {}));
         });
 
-        this.#dataChannel.onClosed(() => {
-            // Simulate closing event
-            if (!this.#closeRequested) {
-                this.#readyState = 'closing';
-                this.dispatchEvent(new Event('closing'));
-            }
 
-            setImmediate(() => {
-                this.#readyState = 'closed';
-                this.dispatchEvent(new Event('close'));
-            });
-        });
+    // we need updated connectionstate, so this is delayed by a single event loop tick
+    // this is fucked and wonky, needs to be made better
+    this.#dataChannel.onClosed(() => setTimeout(() => {
```

Comment:
the spec is very assine about how and when datachannels should close, this is the closest i was able to bring it inline with said spec, the order of events in peer, ice and dc on closing is important for some libraries